app/assets/javascripts/discourse/app/lib/collections.js

```diff
@@ -0,0 +1,17 @@
+/*
+  This returns all items of the left collection
+  except items that are also present in the right collection.
+
+  For example:
+  except(["a", "b", "c", "d"], ["a", "b"]) returns ["c", "d"]
+ */
+export function except(left, right) {
+  const result = [];
+  const rightSet = new Set(right);
+  for (const item of left) {
+    if (!rightSet.has(item)) {
+      result.push(item);
+    }
+  }
+  return result;
+}
```

Comment:
Any opinion on adding this @CvX ?

Reply:
I see it's used to get the difference between array of mentions in a single chat message and an array of loaded users. The first array has the max size of 10 and usually is empty. `a.filter(x => !b.includes(x))` is good enough for that

Reply:
Oh, I actually forgot to remove a restriction here, I wanted to remove a call to `slice`, so instead of 

```javascript
  get #notLoadedMentions() {
    const parsed = parseMentionedUsernames(this._cooked).slice(
      0,
      this.siteSettings.max_mentions_per_chat_message
    );
    const loaded = [...this.mentionedUsers.values()].map((u) => u.username);
    return except(parsed, loaded);
  }
```

we have just

```javascript
  get #notLoadedMentions() {
    const parsed = parseMentionedUsernames(this._cooked);
    const loaded = [...this.mentionedUsers.values()].map((u) => u.username);
    return except(parsed, loaded);
  }
```

This needs an explanation. The thing is that `max_mentions_per_chat_message` is mostly used for restricting amount of users to be notified and that enforced on the server. But if you mention more users than allowed, the mention in the chat message will still look like a mention, so we should show user status on it, otherwise it will be misleading. The default value for this setting is 5, and if I mention 6 users I will be shown a popup that users won't be notified:

<img width="500" alt="Screenshot 2023-06-21 at 17 23 57" src="https://github.com/discourse/discourse/assets/1274517/0fe18728-da7f-4a42-8e89-1905c797c91c">

that's great, but the message will look like this, the sixth user won't have status, while in fact he has one:

<img width="500" alt="Screenshot 2023-06-21 at 17 24 03" src="https://github.com/discourse/discourse/assets/1274517/a017ae58-7682-4a5f-ade6-91303314e766">

So we need to remove a restriction here and pass as many usernames as was mentioned to `except`. From this point of view, looks like it make sense to have this except function instead of using `a.filter(x => !b.includes(x))` that has N^2 complexity?

I've just pushed a commit that removed a restriction. @jjaffeux may you have a look too? Note that this shouldn't lead to problems on the server – it's only one simple query on the server side, and if a user tries to mention a lot of users they either exceed the maximum message size or Rails will decline the request because the query string is too long, otherwise the server should be able to serve the request in reasonable time.

Reply:
We have to make a tricky choice here – show status on all mentions, but in this case sometime we may need to load a lot of users with status, or forbid it, but then status on mentions may be misleading.

Reply:
I think in this case we don't really care about the "slower" perf function given it's called a limited amount of times on a very limited amount of elements, or am I missing something?

Reply:
I've pushed a commit that remove the `except` and uses `a.filter(x => !b.includes(x))` instead.